typescript/xstate/src/lib.ts

```diff
@@ -0,0 +1,360 @@
+import {
+  Actor,
+  ActorLogicFrom,
+  ActorOptions,
+  ActorSystem,
+  ActorSystemInfo,
+  AnyActorRef,
+  AnyEventObject,
+  AnyStateMachine,
+  createActor as createXActor,
+  EventFromLogic,
+  EventObject,
+  HomomorphicOmit,
+  InputFrom,
+  InspectionEvent,
+  InteropSubscribable,
+  Observer,
+  Snapshot,
+  Subscription,
+  toObserver
+} from "xstate";
+import * as restate from "@restatedev/restate-sdk";
+import {TerminalError} from "@restatedev/restate-sdk";
+import {promiseMethods} from "./promise";
+import {
+  AreAllImplementationsAssumedToBeProvided,
+  MissingImplementationsError
+} from "xstate/dist/declarations/src/typegenTypes";
+
+export interface RestateActorSystem<T extends ActorSystemInfo> extends ActorSystem<T> {
+  _bookId: () => string;
+  _register: (sessionId: string, actorRef: AnyActorRef) => string;
+  _unregister: (actorRef: AnyActorRef) => void;
+  _sendInspectionEvent: (
+    event: HomomorphicOmit<InspectionEvent, 'rootId'>
+  ) => void;
+  actor: (sessionId: string) => AnyActorRef | undefined,
+  _set: <K extends keyof T['actors']>(key: K, actorRef: T['actors'][K]) => void;
+  _relay: (
+    source: AnyActorRef | SerialisableActorRef | undefined,
+    target: AnyActorRef,
+    event: AnyEventObject
+  ) => void;
+  api: XStateApi<ActorLogicFrom<T>>
+  ctx: restate.RpcContext,
+  systemName: string,
+}
+
+export type SerialisableActorRef = {
+  id: string;
+  sessionId: string;
+  _parent?: SerialisableActorRef,
+}
+
+export const serialiseActorRef = (actorRef: AnyActorRef): SerialisableActorRef => {
+  return {
+    id: actorRef.id,
+    sessionId: actorRef.sessionId,
+    _parent: actorRef._parent === undefined ? undefined : serialiseActorRef(actorRef._parent)
+  }
+}
+
+type SerialisableScheduledEvent = {
+  id: string;
+  event: EventObject;
+  startedAt: number;
+  delay: number;
+  source: SerialisableActorRef;
+  target: SerialisableActorRef;
+  uuid: string;
+}
+
+async function createSystem<T extends ActorSystemInfo>(ctx: restate.RpcContext, api: XStateApi<ActorLogicFrom<T>>, systemName: string): Promise<RestateActorSystem<T>> {
+  const events = await ctx.get<{ [key: string]: SerialisableScheduledEvent }>("events") ?? {}
+  const childrenByID = await ctx.get<{ [key: string]: SerialisableActorRef }>("children") ?? {}
+
+  let idCounter = 0;
+  const children = new Map<string, AnyActorRef>();
+  const keyedActors = new Map<keyof T['actors'], AnyActorRef | undefined>();
+  const reverseKeyedActors = new WeakMap<AnyActorRef, keyof T['actors']>();
+  const observers = new Set<Observer<InspectionEvent>>();
+
+  const scheduler = {
+    schedule(_source: AnyActorRef, _target: AnyActorRef, event: EventObject, delay: number, id: string | undefined): void {
```

Comment:
scheduler implementation is the first way we can do stuff across time; allows us to delay sending events by using restates `sendDelayed` abstraction. however this is not currently cancellable so we need to keep track of what we have sent and cancelled and check again on arrival